### ISSUE TITLE
Question on intended refactoring and code quality improvements: minimum supported python version?

### ISSUE THREAD

**Issue by benshapiro981** [bleachbit/bleachbit#1631]
**TLDR: interested contributor, want to implement improvements, but have some Qs first** 

Hello!

First of all, big thank you to the maintainers and core devs of this project.  I came across bleachbit a few weeks ago, and it immediately caught my interest. When I looked into it and saw that not only was it a programmable and secure way to clean up files but also that this was THE infamous software which hillary clinton used to delete her emails... you already know I immediately started reading the source code to see if/how I could contribute to the project.

After reading the codebase, I found several ways that I can help out (listed below).  The proposed changes are intended to simplify the build process, make the code easier to work with/maintain/develop, reduce overall codebase size, catch more bugs during static analysis, and potentially speed up the app.  I've been able to implement similar changes on legacy codebases before, and I hope to show that it will be worth it.  At a high level, the improvements are aimed towards....

1. Fixing the structure of the package to reduce the need to do imports inside of function calls (I assume those are currently there to fix circular import errors).
2. Adding [PEP 484-style type hints](https://peps.python.org/pep-0484/) to give editor suggestions for tab-complete and allow tools like [mypy](https://mypy.readthedocs.io/en/stable/) or [pyright](https://microsoft.github.io/pyright/#/) to catch bugs during static analysis.
3. Adding a dependency management system: even something simple like `python -m venv` in combination with a `requirements.txt` file will simplify the setup of the dev and testing environments.  A more reliable and sturdy solution would be to either implement a [PEP 517-style build system](https://peps.python.org/pep-0517/) with something like [meson](https://github.com/mesonbuild/meson) as the actual build backend or to add [conda](https://docs.conda.io/en/latest/) to the repo which can manage both python and non-python dependencies (aka gtk in this case).  This will eventually remove the need for the fragile build scripts made specifically for this project and just allow bleachbit to use well-specced standards.
4. Removing deprecated code, fixing excessively nested/awkward conditionals, using comprehension over for-loops where possible, switching to use of generators (functions with `yield`) over for loops where possible, and other optimizations.
5. IF all of the above 4 are done, I speculate that a reliable and most importantly maintainable **MacOS build with GUI** is possible.  I know that it has been a requested feature for a long time, judging by the website and some of the other issues.

To clarify, I am **ready** and **able** to start submitting pull requests to gradually merge the proposed changes (gotta implement them first).  I understand how important maintaining high coverage % and minimizing regressions is to a project like this, so I indend to make my PRs small and do this in small segments over time.  I have also already read the docs and contribution guide.  However, I have a couple clarifying questions to the current maintainers: 

1. What is the targeted minimum version for python support?  Currently, the earliest version that is not officially end-of-life is 3.8.  However, in this repo I see some code suitable for versions which are much older.  The features I suggested above would work for 3.6 and up.
2. What are the near term milestones and existing plans (if any) for development and new features?  I want to work with you around your existing goals as much as possible and not cause confusion.  

Thank you for taking the time to read through this!  And again thank you for your work in maintaining this project.

**Comment by rados** [comment]
@Up50qyx Hey, great thanks! I am Rado, my role here is just supportive. In my humble opinion it would be excellent if we could increase the test coverage and then we will be more confident to refactor...

**Comment by az0** [comment]
Thank you for your patience with my late reply

> big thank you to the maintainers and core devs of this project.

You're welcome! Thanks to you for the interest in making improvements!

> Fixing the structure of the package to reduce the need to do imports inside of function calls (I assume those are currently there to fix circular import errors).

I would approach this carefully. Reasons include

- Avoid circular import (as you mentioned)
- Lazy import for speed (until we can rely on native lazy imports in Python 3.15)
- Allow some features/packages to be optional (e.g., FreeBSD has separate package for sqlite3) without crashing
- Separate builds for wxPython, GTK, and TUI (e.g., do not require both wxPython and GTK in one Windows build)

> Adding [PEP 484-style type hints](https://peps.python.org/pep-0484/) to give editor suggestions for tab-complete and allow tools like [mypy](https://mypy.readthedocs.io/en/stable/) or [pyright](https://microsoft.github.io/pyright/#/) to catch bugs during static analysis.

OK

> Adding a dependency management system: 

I'm open to discussion. First, take a look at recent changes, which include

- requirements.txt, requirements-test.txt, windows/requirements.txt
- `make install-deps` and - `make install-deps-dev`
- `scripts/install-deps.sh

I also want to keep it easy for new developers to get started without complex environmental setup or heavy requirements.

> Removing deprecated code, fixing excessively nested/awkward conditionals, using comprehension over for-loops where possible, switching to use of generators (functions with yield) over for loops where possible, and other optimizations.

Recently XhmikosR made many such changes, but there is always room for improvements. PRs are welcome, and smaller PRs are easier to review and merge than big PRs.

(Sorry, the same principle applies to issues, discussions, questions: it's easier to read and to reply to a single, focused question than one issue with many questions.)

> so I indend to make my PRs small and do this in small segments over time. 

Perfect!

>IF all of the above 4 are done, I speculate that a reliable and most importantly maintainable MacOS build with GUI is possible. I know that it has been a requested feature for a long time, judging by the website and some of the other issues.

Recently there's been progress. See https://github.com/bleachbit/bleachbit/issues/1863#issuecomment-5553548302 for a recent update

>What is the targeted minimum version for python support? 

According to the [official Python status page](https://devguide.python.org/versions/), the EOL dates

- Python 3.8 : October 2024
- Python 3.9 : October 2025
- Python 3.10  : October 2026

However, vendors may extend support.

Current code requires Python 3.8, but based on review, that can be increased to 3.9

Minimum Python version by OS

- Windows: Python 3.12
- macOS: Python 3.9.6 (easily accessible with Apple Xcode)
- Linux: 3.9 now ( could be 3.10)

Quick review of supported Linux distributions

- Ubuntu 22.04 LTS has Python 3.10 and is EOL April 2027
- Debian 11 has Python 3.9 and is EOL August 2026. Today I disabled it.
- Debian 12 LTS has Python 3.11 and is EOL date June 2028
- RHEL/Alma 9 has Python 3.9, full support ends May 2027, maintenance support ends 2032
- RHEL/Alma 10 has Python 3.12 

I would consider dropping RHEL/Alma 9, so that brings the minimum version to Python 3.10. I'd also consider requiring Python 3.10 on macOS.

If there are particular changes between Python 3.9 and 3.10 that are important, let me know.

> What are the near term milestones and existing plans (if any) for development and new features? I want to work with you around your existing goals as much as possible and not cause confusion.

6.0.4 is a maintenance release with countless bug fixes, refactors, security fixes, code cleanup. It also improves macOS support (but no DMG or beta quality).

Near-term priorities include:

- First macOS beta with GUI and .dmg
- Separate cleaning worker vs GUI for responsive GUI on all platforms. Also, run GUI on Wayland/X11 as non-elevated process.
- Refactor large modules
- Replace markovify with faster code (ready to go)
- Send deleted files to trash (vs standard delete or overwriting)
- Automatically close running browser
- Keep new files in trash, recycle bin, temp folder, and maybe other paths
- Integrate TUI and wxPython branches to master
- Simple GUI option for novice users
- Add launcher on Windows (to address ~10 distinct issues)
- Add more, high-demand cleaners
- Make it easier for users to send debugging info
- Improve feedback to users when elevated privileges are needed
- Stable release for 64-bit Windows. (Currently TUI, CLI, or wxPython are 64-bit on Windows, but Windows GTK is 32-bit.)
- Review bleachbit/bleachbit and bleachbit/issue for other features that users want. (You can look for duplicate requests, issues with lots of discussion or reactions, or the "important" label.)
